bin/k8s-deploy

```diff
@@ -232,6 +231,96 @@ do
+for index in "${!BG_DEPLOYMENT_FILES[@]}"
+do
+  DEPLOYMENT=${BG_DEPLOYMENTS[$index]##*/}
+  DEPLOYMENT_FILE=${BG_DEPLOYMENT_FILES[$index]}
+  if [ -n "$CI_SHA1" ]
+  then
+    cp ${DEPLOYMENT_FILE} ${DEPLOYMENT_FILE}-tmp
+    echo "Updating image tag from latest to ${CI_SHA1} for ${DEPLOYMENT_FILE}"
+    sed 's/:latest/':${CI_SHA1}'/g;' ${DEPLOYMENT_FILE}-tmp > ${DEPLOYMENT_FILE}-${CI_SHA1}
+    mv ${DEPLOYMENT_FILE}-${CI_SHA1} ${DEPLOYMENT_FILE}-tmp
+
+    echo "Updating deployment label with version: ${CI_SHA1} for ${DEPLOYMENT_FILE}"
+    sed 's/version: latest/version: '${CI_SHA1}'/g;' ${DEPLOYMENT_FILE}-tmp > ${DEPLOYMENT_FILE}-${CI_SHA1}
+    mv ${DEPLOYMENT_FILE}-${CI_SHA1} ${DEPLOYMENT_FILE}-tmp
+
+    echo "Updating deployment name with version: ${CI_SHA1} for ${DEPLOYMENT_FILE}"
+    sed '/name:.*-latest/s/latest/'${CI_SHA1}'/g' ${DEPLOYMENT_FILE}-tmp > ${DEPLOYMENT_FILE}-${CI_SHA1}
+    
+    rm ${DEPLOYMENT_FILE}-tmp
+    
+    DEPLOYMENT_FILE=${DEPLOYMENT_FILE}-${CI_SHA1}
+
+  fi
+
+  # This is workaround attended to address an HPA issue where the correct number of pods are not deployed when
+  # 'kubectl apply' is used to redeploy a deployment that has been scaled by an HPA.  There is an open issue for
+  # this: https://github.com/kubernetes/kubernetes/issues/25238
+
+  # If we can match 'replicas: hpa' in te deployment file
+  if [[ -n $( awk -F: '/replicas: hpa/' $DEPLOYMENT_FILE ) ]]
+  then
+    # Check to see if the deployment exists yet. If it does then get current replicas, if not set to default.
+    if [[ ! -z $(kubectl get deployments --namespace=${NAMESPACE} -o json | jq -r '.items[].metadata.name' | grep -x ${DEPLOYMENT}) ]]
+    then
+      REPLICA_COUNT=$(kubectl get deployment ${DEPLOYMENT} --namespace=${NAMESPACE} -o json | jq '.status.readyReplicas')
+    else
+      REPLICA_COUNT=${DEFAULT_REPLICA_COUNT}
+    fi
+    # Sed replica count into the deployment file in place of 'hpa'
+    echo "Updating replica count to ${REPLICA_COUNT} for ${DEPLOYMENT_FILE}"
+    sed -e 's/replicas: hpa/replicas: '${REPLICA_COUNT}'/gI;' ${DEPLOYMENT_FILE} > ${DEPLOYMENT_FILE}-hpa
+    DEPLOYMENT_FILE=${DEPLOYMENT_FILE}-hpa
+  fi
+
+  kubectl apply -f ${DEPLOYMENT_FILE} --namespace=$NAMESPACE --record
+  DEPLOYMENT_NAME=$(cat ${DEPLOYMENT_FILE} | shyaml get-value metadata | grep name | awk '{ print $2 }') 
+  
+  # Wait until the Deployment is ready by checking the MinimumReplicasAvailable condition.
+  READY=$(kubectl -n $NAMESPACE get deploy $DEPLOYMENT_NAME -o json | jq '.status.conditions[] | select(.reason == "MinimumReplicasAvailable") | .status' | tr -d '"')
+  while [[ "$READY" != "True" ]]; do
+    READY=$(kubectl -n $NAMESPACE get deploy $DEPLOYMENT_NAME -o json | jq '.status.conditions[] | select(.reason == "MinimumReplicasAvailable") | .status' | tr -d '"')
+    sleep 5
+  done
+
+  echo "Updating Service for Blue Green Deployment"
+  for index in "${!SERVICE_FILES[@]}"
+  do
+    SERVICE_FILE=${SERVICE_FILES[$index]}
```

Comment:
I don't think this is a universally safe assumption to make. Services:Deployments is a many:many relationship.

Reply:
If we can't detect it, we'll have to be explicit with a new config field. Since BG only matters if there is an existing deployment, maybe it can be more easily dynamically pulled from the k8s api?

Reply:
Does the API have more information that the YAML files will in this case? 
This seems like a thing that is easy to to manually, but hard to do automatically if we make it as flexible as you are thinking is could/should be. Would it be safer to set a smaller goal and some rules and narrow the scope? Maybe define the service name in the deployment as an annotation?